server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/restore/TransportRestoreSnapshotAction.java

```diff
@@ -72,7 +73,11 @@ protected void masterOperation(
     ) {
         restoreService.restoreSnapshot(request, listener.delegateFailure((delegatedListener, restoreCompletionResponse) -> {
             if (restoreCompletionResponse.getRestoreInfo() == null && request.waitForCompletion()) {
-                RestoreClusterStateListener.createAndRegisterListener(clusterService, restoreCompletionResponse, delegatedListener);
+                RestoreClusterStateListener.createAndRegisterListener(
+                    clusterService,
+                    restoreCompletionResponse,
+                    new ContextPreservingActionListener<>(threadPool.getThreadContext().newRestorableContext(true), delegatedListener)
```

Comment:
I think it'd be better to capture the context deeper down. Perhaps within the constructor to `RestoreClusterStateListener` (solving two problems addressed in this PR), but actually maybe the `ClusterApplierService` should capture the context on all listeners that get added?

For comparison, the `MasterService` takes responsibility for completing master tasks' listeners in the right context.

Reply:
I pushed the preservation down to the ClusterApplierService when saving the listener to the list of available listeners.